### ISSUE TITLE
Tracking issue for format_args_capture

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#67984]
This is a tracking issue for the RFC 2795 (rust-lang/rfcs#2795).

**Steps:**

[stabilization-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs
Checklist for the tracking issue:

Steps:
- [x] Implementation for `format_args!` behind `format_implicit_args` feature (name to be bikeshed?)
- [ ] Implement alternative solution for `format_args!(concat!(...))` case that doesn't risk [spurious macro hygiene](https://github.com/rust-lang/rfcs/pull/2795#issuecomment-550403099)
- [x] Implementation for `panic!`. Behind feature gate and / or edition switch (see ~~un~~resolved question) #80162
- [ ] Adjust documentation ([see instructions on rustc-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-guide][stabilization-guide])

Unresolved Questions:
- [x] Solution for `format_args!(concat!(...))` - perhaps try out a new unstable macro as per [this comment](https://github.com/rust-lang/rfcs/pull/2795#issuecomment-559245529) - Resolved by https://github.com/rust-lang/rust/issues/67984#issuecomment-801394736
- [x] Final design of the `panic!` solution - Solved by [rfc 3007](https://github.com/rust-lang/rfcs/pull/3007)

**Comment by davidhewitt** [comment]
Progress update from me: I've been busy but have started work on the implementation of this feature.

A bikeshed on the feature name: I'm leaning towards to the name `format_args_capture` over the original proposed name `format_args_implicits` from the RFC.

I think that `forward_args_capture` reflects the mechanics of the feature better - it enables capturing of variables from the surrounding scope. It also avoids the word "implicit" which I find can raise peoples' blood temperature (sometimes for good reason).

Is it too late to use a different feature name?

**Comment by programmerjake** [comment]
maybe `format_args_capture_shorthand`? "shorthand" is used in the docs for [`struct` field init shorthand](https://doc.rust-lang.org/reference/expressions/struct-expr.html#struct-field-init-shorthand).

**Comment by joshtriplett** [comment]
@davidhewitt `format_args_capture` seems reasonable, given the error message, and I agree with avoiding the word `implicit`.

**Comment by Mark-Simulacrum** [comment]
FWIW at least in my mind capturing implies something closer to what e.g. closures do where the variable is stored within the returned Arguments struct. But that's not something that can be added to Arguments given our stability guarantees, I think, even though I've often wanted it -- so maybe that's not a major concern.

I will note though that the naming of the feature gate and diagnostics feel like something that we can adjust over time and in the implementation PR, definitely no need to get them right immediately :)

**Comment by m-ou-se** [comment]
>   * [ ] Final design of the `panic!` solution - perhaps based off [this comment](https://github.com/rust-lang/rfcs/pull/2795#issuecomment-559246909)

Some notes about this problem:

---

There's several macros that currently that handle their no-formatting-arguments case specially, mostly for efficiency:

- `std::panic!()`
- `core::panic!()`
- `assert!()`, which forwards to either `std::panic!()` or `core::panic!()`
- all kind of log/panic/etc. macros in the rust ecosystem, see [this comment](https://github.com/rust-lang/rfcs/pull/2795#issuecomment-559219291).

In all cases, this results in these two problems:

1. `x!("{}")` doesn't give an error, but just prints/uses literal `"{}"`.
2. `x!("{var}")` will not work with this RFC.

In the case of `std::panic!()`, this happens because it accepts objects of any type. E.g. `panic!(5)` will 'throw' a `Box<Any>` containing `5`. So a `panic!("hello")` will trigger that same case.

`core::panic!()` doesn't throw boxes around, but still has a special case for efficiency. `core::panic!("hello")` will call `core::panicking::panic` (instead of `panic_fmt`), which takes a `&str` and constructs an optimal `fmt::Arguments` using that `&str` to avoid pulling in any formatting code.

Most of the crates from the community using a special no-arguments case are doing this for efficiency as well. Some will use `args.to_string()` to format the message into a buffer first, to reduce the number of write/log/etc. calls. This would be an unnecessary copy of a `&str` in the trivial case, which they try to avoid.

---

In the case of `core::panic` and the macros found in various crates, this *could've been* solved if [`Arguments::as_str() -> Option<&str>`](https://github.com/rust-lang/rust/pull/74056) had existed, which would've meant that none of these macros needed a special case for efficiency. The trivial case could've been handled by the printing code instead, and the compiler would've easily eliminated the allocating case completely from the resulting binary. Unfortunately, changing that now will be a breaking change.

Not only will `x!("{}")` stop working, but most of these macros (including `core::panic!()`) also accept full expressions that evaluate to a `&str` right now, because they simply forward their one argument to a function that takes a `&str`. So:

- `core::panic!(&String::from("hello"))` is accepted today.

---

Unless breaking these examples is acceptable, this would need to be changed over a Rust edition. For both `std`'s and `core`'s `panic!()` (and `assert!()`). What exact behaviour makes most sense (also in terms for warnings or clippy lints) in edition 2018 at that point, is still unclear. What behaviour makes sense in edition 2021 is also not clear:

- What do do with `std::panic!(&String::from("hello {}"))`? If it accepts arbitrary objects, this would still compile, but act differently than `std::panic!("hello {}")`. Also, special casing on a string literal is currently not possible with `macro_rules!()`. `$x:literal` is possible, but would break `panic!(5)`. Perhaps a separate panic function or macro for throwing boxes is necessary. (In that case, using `panic!()` for arbitraty objects could already get deprecated in edition 2018.)
- Same question for `core::panic!()`. The most trivial solution would mean that `core::panic!({ let x = "abc"; x })` no longer compiles. Special casing on `$x:literal` would work fine here, but it might be better to keep things simple.
- (`assert!()` follows the behaviour of either `core::panic!()` or `std::panic!()` depending on whether it is used in `no_std`.)

For the macros in the community crates, there are two options w.r.t. an edition change:

1. Expose some macro like `format_args_or_literal!()` to allow them to have the same behaviour over editions as the core panic macro. This macro would either return an enum, or expect users to use `.as_str()` as described above to handle the trivial case. All these crates will need to be updated, but their behaviour will not change for users staying on edition 2018.
2. Expect crates to update to using `.as_str()` to handle the trivial case, and not tie their behaviour to the Rust edition. These crates will have to release a breaking change and bump their version. (Resulting in a bit more breakage but less complexity.)

---

My proposal:

- Update *all* macros to always use `format_args!()`, without the special case, but the core/std macros only in edition 2021. This means dropping support for `panic!("{}")`, `panic!({ let x = "abc"; x})`, `panic!(5)`, `panic!(bla)` etc.
- Use `Arguments::as_str()` for optimizing the trivial case, instead. Expect community crates to start using that in all editions.
- Add a separate `panic()` function in `std` to throw `Box<Any>`s.

**Comment by nikomatsakis** [comment]
Thanks @m-ou-se for the detailed write-up. I'm nominating this for @rust-lang/lang discussion since it has Edition ramifications, although I guess that the design of these macros is in many ways a @rust-lang/libs decision. 

I agree with you that an edition is the appropriate way to manage this, and I think we should shoot for `panic!("{var}")` working consistently. We'd have to look carefully at how to implement this, I don't think it would work using the mechanisms we have, but I think it could be made to work. I'd probably say that "non-string-constants" should just be deprecated, but I don't have a strong opinion about it. 

The automated migration seems easy enough, in any case, since `panic!(...)` just becomes `panic!("{}", ...)`, right?

**Comment by m-ou-se** [comment]
> The automated migration seems easy enough, in any case, since `panic!(...)` just becomes `panic!("{}", ...)`, right?

For `core::panic!()`, yes. But for std's, no. `std::panic!("hello")` throws a `&str`, not a `fmt::Arguments` or `String`. The standard panic handler would print the same, but custom panic hooks can tell the difference. Also, `std::panic!()` can throw things that don't implement Display/Debug right now.

**Comment by Amanieu** [comment]
I can see the `panic!` change working by exporting a different macro in the new 2021 prelude.

I'm generally in favor of removing `Box<Any>` support from the `panic!` macro: it's very rarely used and as @m-ou-se said, this functionality can still be provided via a normal function in std.

**Comment by m-ou-se** [comment]
Something missing from my notes/proposal above is making the behaviour consistent or at least not too unexpected on 2018/2015, after `format_implicit_args` becomes stable.

If the changes to core/std's panic!() are gated behind edition 2021, then in 2018/2015 we get:

```rust
let a = 5;

println!("{a}"); // prints "5"
panic!("{a}"); // panics with "{a}"

println!("{a} {}", 5); // prints "5 5"
panic!("{a} {}", 5); // panics with "5 5"
```

Possible ways to make this better:
1. Gate all of `format_implicit_args` behind edition 2021. No `println!("{a}")` in 2018. :(
2. Gate `format_implicit_args` for panics behind edition 2021. No `panic!("{a} {}", 5)` in 2018.
3. Warn for 1-argument panics containing implicit arguments. `panic!("{a}")` warns, but still prints `"{a}"`. (But what should the warning suggest? `panic!("{}", a)`? `v2::panic!("{a}")`? "Switch to 2021"? ..?)
4. Break things and don't gate anything behind an edition, and hope that literal braces (and non string literals) in single argument panics are uncommon right now. Possibly by starting giving deprecation warnings first for a while. (Though what should those warnings suggest for `panic!("{")`? `panic!("{}", "{")`? `panic_str("{")`? ..?)

None of these seem great. :(

**Comment by jplatte** [comment]
>     4. Break things and don't gate anything behind an edition, and hope that literal braces (and non string literals) in single argument panics are uncommon right now. Possibly by starting giving deprecation warnings first for a while. (Though what should those warnings suggest for `panic!("{")`? `panic!("{}", "{")`? `panic_str("{")`? ..?)

This could also be `panic!({ "{" })` / `panic!(("{"))`, right? Still not great of course.

**Comment by m-ou-se** [comment]
> This could also be `panic!({ "{" })` / `panic!(("{"))`, right? Still not great of course.

Maybe, but *if* `panic!("{")` would be deprecated, I suppose using expressions like that should be deprecated as well.

**Comment by nikomatsakis** [comment]
We discussed this in today's @rust-lang/lang meeting. The sense of the meeting was:

* We would be in favor of completing this RFC and using an edition transition to make `println!`, `panic!`, etc all behave in a uniform fashion.
* We would like to do so in a way however that preserves all existing capabilities (e.g., ability to panic with arbitrary values, and with string literals without allocation), though that may require more explicit syntax. We did not discuss in detail the comments I see here, such as whether `panic!({"{"})` ought to be deprecated, however.
* We would like to see a concrete proposal with a migration plan.

**Comment by nikomatsakis** [comment]
> We did not discuss in detail the comments I see here, such as whether `panic!({"{"})` ought to be deprecated, however.

My personal opinion is that `panic!({expression})` would be quite reasonable as a way to 'panic with some explicit value', as opposed to `panic!("...", ...)` as the "with arguments" form.

One interesting question that I don't know the answer to is whether `panic!("foo")` ought to panic with a `&'static str` or a `String` -- it introduces an obvious "non-uniformity" in the type, depending on whether it contains "format notation". I do think that `panic!("foo{{")` ought to be required in order to name a `{` (i.e., we should interpret a string literal as a format string). I think I would tend towards `panic!("foo")` raising a `String`, and requiring `panic!({"foo"})` to panic specifically with a string literal, but that's a lightly held opinion.

**Comment by Amanieu** [comment]
My preference would be to deprecate `panic!` with arbitrary expressions and only support format strings. I have never seen a use of non-string panics in the wild, so I think this functionality is better provided by a function in `std::panicking` that takes a `Box<Any>`.

**Comment by Lokathor** [comment]
Supporting Anicdote: I used a non-string panic once on accident and it took me well over 30 seconds to even understand what had gone wrong (because the compiler accepted it, so "clearly" i must have typed the right thing).

**Comment by davidhewitt** [comment]
I too have only only ever experienced non-string panics in error. Especially the first time I did this it took quite a while to figure out where the `Box<dyn Any>` was originating from.

@nikomatsakis when you say the lang team intends to keep all existing functionality, do you mean that all of the functionality remains on the macros? Or is moving the "panic with expression" case to a new function potentially on the table?

My reservation with the proposed `panic!({foo})` syntax for "panic with box any containing foo" is that it's dangerously close to `panic!("{foo}")`. I think especially beginner Rust users won't easily spot the difference and will also be the ones most confused when they find their panic message is broken.

As for a migration strategy, I prefer @m-ou-se 's option 3. I would like to propose that we do the following things at the same time without waiting for an editon change:
- we introduce this RFC for all cases except single-argument `panic!`, as this RFC is non-breaking for these cases.
- we introduce the new syntax for "panic with a value", whether function or new syntax in the `panic!` macro
- for the single-argument panic, we start to emit a warning for string literals which contain format syntax, and otherwise continue to let it function as-is. The warning could suggest two options of resolution which would not change after migration to the next edition. E.g. for `panic!("{foo}")` - this would be `panic!(format!("{foo}"))` if fornatting was intended, and maybe `panic!({"{foo}"})` if a literal was intended. (Though this nested set of sigils is quite heavy, which again makes me think migrating the "panic with a value" case to a function would be desirable).

After the edition change the single-argument panic case changes to exclusively expect a format string, and we're done!

I think this can be implemented by changing the single-argument panic arm to forward to a new compiler-builtin macro. This macro would be the only code that needs to change behaviour with an edition change.

I need to think further about the exact design of this new built-in macro. I think some downstream libraries which behave similarly to `panic!` may also want to use this macro to support cross-edition behaviour. I'll try to come back with a design hopefully at the weekend if y'all haven't already figured it out by then! 😄

**Comment by nikomatsakis** [comment]
I agree that `panic!({"foo"})` is perhaps too close to `panic!("{foo}")`. Another option is to have a `panic_with!` macro or something like that (though this name is perhaps non-ideal, since it suggests a closure).

> I would like to propose that we do the following things at the same time without waiting for an editon change:

In general, I think the way to "phase in" new syntax should fit this mold (and your suggestion does):

* Permit the new syntax when we can do so backwards compatibly.
* Have an "uglier" form that you can use until the next edition (`panic!("{foo}")`, in your proposal).
* Introduce the "cleaner" form after new edition.

I guess my one concern with the plan as described is that it seems like it's a bit surprising that `panic!("{x} {}", "foo")` will print out `x` but `panic!("{x}")` will not. It might be cleaner to have folks "opt in" to the new semantics completely, perhaps by an explicit `use std::prelude::v2021::panic` or something like that? Have to think about that a bit.

**Comment by davidhewitt** [comment]
> Another option is to have a panic_with! macro or something like that (though this name is perhaps non-ideal, since it suggests a closure).

`panic_with!` is actually something I was wondering about yesterday too, though I hadn't considered the point you make about the closure.

One idea that crossed my mind was simply making this panic-with-a-value case a function called `start_unwind`. It fits well with the existing `catch_unwind` / `resume_unwind`. OTOH if it needs to stay a macro, I might avoid `start_unwind!` *because* the other two are functions.

I kind of like `throw!` or `panic_throwing!` (because it pairs with "catch" in `catch_unwind!`). Maybe of the two, `panic_throwing!` is better because it keeps the `panic` terminology rather than suggesting this is some alternate error handling mechanism. 

> I guess my one concern with the plan as described is that it seems like it's a bit surprising that panic!("{x} {}", "foo") will print out x but panic!("{x}") will not. It might be cleaner to have folks "opt in" to the new semantics completely, perhaps by an explicit use std::prelude::v2021::panic or something like that? Have to think about that a bit.

👍 I don't have a strong opinion on whether it should be opt-in so happy to defer to core team members on this point.

One thing to consider is that perhaps it's already surprising for a lot of people that `panic!("{x}")` is not a compile error - so by starting to emit a warning for this case we might rescue a few people from hidden bugs!

**Comment by Lokathor** [comment]
`panic!` is a macro because it does formatting.

Presumably if it's taking a single value the `start_unwind` operation could just be a generic function like `fn start_unwind<T>(t: T) -> !`

**Comment by Nemo157** [comment]
Panicking doesn't necessarily unwind, with the `panic=abort` setting, or with `no-std` and a custom `panic_handler`.

**Comment by davidhewitt** [comment]
> Panicking doesn't necessarily unwind, with the panic=abort setting, or with no-std and a custom panic_handler.

Great point - in which case I think `start_unwind` is not a good name for this any more!

**Comment by m-ou-se** [comment]
Opened #74622 specifically for this arbitrary-payload-throwing-panic function, to have a place to bikeshed over the name etc.

**Comment by RalfJung** [comment]
Regarding the `panic!` changes, even the already performed changes to `core::panic!(foo)` (as part of https://github.com/rust-lang/rust/pull/74056) are potentially problematic:
* they break some cases of panics in consts (https://github.com/rust-lang/rust/issues/51999#issuecomment-687590082) (a nightly-only regression)
* they lead to increased code size by circumventing [the trick explained here](https://github.com/rust-lang/rust/blob/c3364780d2cfddfe329f62a3ec138fd4f9a60e27/library/core/src/panicking.rs#L44-L50).

So whatever we end up doing should be done uniformly for both of these macros -- and maybe there should be some discussion if the above mentioned regressions are acceptable given that there is not even a clear transition plan yet.

**Comment by RalfJung** [comment]
I forgot to say, both issues could be avoided if `Arguments::as_str` returned `&str` instead of `&'static str` -- but that might cause other problems elsewhere. This was mentioned at https://github.com/rust-lang/rust/pull/74056#issuecomment-653896585 but I do not understand how the static lifetime makes such a big difference here -- `Cow<'_, str>` also works very well. On the one hand I can see the argument that format strings are static anyway, on the other hand it does seem potentially useful for `Arguments` to be able to support allocation-free non-static strings as well.

**Comment by m-ou-se** [comment]
> I do not understand how the static lifetime makes such a big difference here

`as_str()` was made to avoid the need for special-casing macros (like `bail!()` [here](https://github.com/rust-lang-nursery/failure/blob/20f9a9e223b7cd71aed541d050cc73a747fc00c4/src/macros.rs#L9-L17), or like `anyhow::anyhow!()`, etc.). They can only use it if it is `'static`, as they keep the string around in an object, instead of printing it directly.

**Comment by RalfJung** [comment]
That makes sense. And this then implies that `core::panic!(foo)` cannot use `core::panicking::panic` any more as `foo` might not be static, which directly leads to the issues I mentioned above.

I don't see a good way out of this, but I am satisfied with having understood the tradeoff here.

**Comment by davidhewitt** [comment]
Relevant discussion regarding panic has recently been proposed in https://github.com/rust-lang/rfcs/pull/3007

**Comment by mbartlett21** [comment]
Are the named arguments meant to work with raw strings?

At the moment, it says that `format_args!` cannot capture variables when the format string is expanded from a macro.

[Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=59dc2e1e9f7e0e9b7897b4c8dc8aa5ed)

**Comment by davidhewitt** [comment]
Looks like an oversight by me - I'll try to put out a PR to fix it within the next couple of days.

**Comment by davidhewitt** [comment]
@m-ou-se I see you removed this issue from "Proposed in 2021 Edition" - did I miss a decision that this feature is not going to be a part of the 2021 edition?

If the decision is still yet to be made, I'd be willing to put the work in to get this feature in.

**Comment by m-ou-se** [comment]
This doesn't need to happen over an edition change, right?

The only thing that needed an edition change was the solution for the panic and assert macros, and that one is already approved and split to a separate issue: https://github.com/rust-lang/rust/issues/80162

**Comment by davidhewitt** [comment]
Ah yes, very true!

**Comment by m-ou-se** [comment]
What's the status of this? What needs to be done before we can stablize this?

Is the checklist in the top comment still correct?

**Comment by davidhewitt** [comment]
As far as I'm aware, no documentation has yet been written for this, which would be the main blocker for stabilization. I imagine that [the documentation for `std::fmt`](https://doc.rust-lang.org/std/fmt/index.html) would be the place which might want the most changes.

The only other unresolved question from above is the `format_args!(concat!(...))` case, which was called out specifically as a common case where the format string is not a literal, but users still may want some form of named argument capture.

In my own reflections on this topic since the RFC was written, my personal preference is to just use nested `format_args!`, e.g.:

```rust
format_args!(concat!(expr, "\n"), arg1, arg2);
```

can be rewritten without "concat" as

```rust
format_args!("{}\n", format_args!(expr, arg1, arg2));
```

This probably introduces a little bit of extra work for the compiler / optimizer, however I think it's probably good enough such that a more complex solution isn't needed.

See for example this playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a3138168fe37bef183e2428cbb8bb8e4

**Comment by davidhewitt** [comment]
(To clarify; I've been very busy recently so I don't think that I'm going to be writing documentation for this any time soon. Anyone who wants to see this reach stable is encouraged to contribute docs!)

**Comment by m-ou-se** [comment]
> As far as I'm aware, no documentation has yet been written for this, which would be the main blocker for stabilization. I imagine that the documentation for std::fmt would be the place which might want the most changes.

Documentation should probably land in the same PR in which this feature gets stabilized. To make sure we don't document things that don't exist yet.

> This probably introduces a little bit of extra work for the compiler / optimizer, however I think it's probably good enough such that a more complex solution isn't needed.

Yeah i mentioned that idea in https://github.com/rust-lang/rust/issues/78356 a while ago, as something we should optimize nicely by 'inlining' nested format_args into eachother. It's far from trivial, but there's now quite a few cases that would benefit from it.

**Comment by m-ou-se** [comment]
@nikomatsakis Do you think this is ready for stabilization as it is? Or are there any issues left that need solving first?

**Comment by nikomatsakis** [comment]
@m-ou-se I'm not aware of any problems remaining! I'd be in favor of a PR with a stabilization report. Ideally that would include a summary of the changes in behavior and links to relevant tests.

**Comment by mitsuhiko** [comment]
I was wondering about an aspect of this RFC which I did not understand if a conversation covered it already. Today with `format_args!` we have a macro which takes a format string + arguments and returns an object which can be passed into a formatting function. Given where this RFC is going and what inspires it, in a way one could imagine that `format_args!("foo {bar}")` might eventually get syntax support in the form of `f"foo {bar}"`.

If that were to happen some patterns that exist currently would be quite odd though. Eg one would have to do `println!("{}", f"foo {bar}")` instead of the more common `println!(f"foo {bar}")`. But since `f"foo"` would expand to `format_args!("foo")` it would create a different type than a bare string literal which could be exploited.

This potential evolution would mean that `println!("{foo}")` might not happen, but `println!("{}", format_args!("{foo}"))` or with some specialization magic `println!(format_args!("{foo}"))` could happen which with a literal would turn into `println!(f"{foo}")`.

**Comment by bstrie** [comment]
> one would have to do println!("{}", f"foo {bar}") instead of the more common println!(f"foo {bar}")

I don't see any technical reason why `println!(f"{bar}")` would be impossible in a hypothetical f-string future. As implemented `println!` already does nothing other than forward a raw token tree on to `format_args!`, which is a magical compiler built-in that actually does all the validation; this token tree would be enough for `format_args!` to distinguish between string literals and f-strings (and accept both).

Furthermore, we could also note that first-class f-strings would remove all reason for `println` to be a macro at all; with the varargs and validation being taken care of by the f-strings themselves, you could just provide a function:
```rust
println(f"{foo}")
```

**Comment by mitsuhiko** [comment]
> I don't see any technical reason why println!(f"{bar}") would be impossible in a hypothetical f-string future.

Neither do I. Sorry my comment was unclear. The ambiguity only exists for `panic!`. I have updated my comment accordingly.

**Comment by m-ou-se** [comment]
@mitsuhiko The panic issue is [already resolved in Rust 2021](https://github.com/rust-lang/rust/issues/80162). `panic!(anything)` in Rust 2021 unconditionally expands to an expression with `format_args!(anything)`:
https://github.com/rust-lang/rust/blob/f811f14006fa46030f1af714f7d640580d3ad822/library/core/src/panic.rs#L33-L40

If we add `f".."` syntax, we should just make `format_args!()` accept those as well as the format string.

This reminds me that we might already want to reserve the `f".."` syntax in edition 2021. I've added it to the agenda for the next edition meeting.

**Comment by davidhewitt** [comment]
Some uncertainty I have about  `f".."` syntax is that if we were going for something semantically equivalent to Python, `f".."` should expand to `format!("..")` rather than `format_args!("..")`.

Expanding to `format!("..")` doesn't look like it would compose nicely with the other macros.

If `f".."` expanded to `format_args!("..")` then `f"{foo}".to_string()` might become a common way to format strings, for symmetry with `"str literal".to_string()`.

----

An alternative worth considering regarding `#[feature(format_args_capture)]` in particular is that f-string syntax could be used to make this feature opt-in.

For example, `format!(f"{foo}")` could try to capture `foo` from surrounding scope, and we could leave `format!("{foo}")` as a compile error (missing named parameter `foo` to the format macro).

If people wanted this feature to be opt-in then I could support this position. My opinion is that as stabilizing this feature won't break any existing code, there's no need to make it opt-in via a new syntax. New syntax means more confusion for beginner Rust learners, and that's one thing that this feature was intended to reduce (by making formatting macros slightly cleaner).

One upside of having opt-in syntax would be that the `concat!` case will then explicitly not combine with with this feature. At the moment this feature states that `format!(concat!("{foo}", "{bar}"))` will be a compile error (format args capture is disabled when the formatting string originates from a macro invocation that expands to a string literal).

Overall, I still defend that this feature as implemented makes the correct compromise between the design points above.

**Comment by bstrie** [comment]
There's enough unexplored design space around first-class f-strings that I think we should avoid bogging down this issue with that discussion. It seems completely reasonable to consider stabilizing this feature as-is, while leaving broader f-string design to a future RFC.

**Comment by dbsxdbsx** [comment]
Once this feature is added, I think there is no need to use this[ crate ](https://github.com/danielhenrymantilla/fstrings-rs) anymore.

**Comment by bstrie** [comment]
So to clarify, the step "Implement alternative solution for format_args!(concat!(...)) case that doesn't risk spurious macro hygiene" has been deemed unnecessary, yes? I'm happy to write up documentation for this if that's the next step.

**Comment by davidhewitt** [comment]
> So to clarify, the step "Implement alternative solution for format_args!(concat!(...)) case that doesn't risk spurious macro hygiene" has been deemed unnecessary, yes?

At least in my opinion, using nested format args as a workaround seems reasonable enough that we don't need to add anything else at the same time as stabilizing this feature.

Please feel free to own this @bstrie - unfortunately I'm too time-poor to push this forward myself at current.

**Comment by yerlaser** [comment]
Sorry for hijacking this discussion, but could you also consider shortening of the format! macro to fmt! or even f!.
This will make using it much simpler in cases like:
println(fmt!"Hello \{name\}")
or
println(f!"Hello \{name\}")

**Comment by davidhewitt** [comment]
@yerlaser you don't need to combine `format!` with `println! ... with this feature you would just write `println!("Hello {name}")`, and on current stable you can do `println!("Hello {}", name)`

**Comment by yerlaser** [comment]
That is good, indeed. I just saw above some mentions of using a println function instead of macro.
However, a shortened format! macro can still be useful in cases like writing to stderr or a file, etc.

**Comment by RalfJung** [comment]
@yerlaser please check our API docs, we do have macros for that as well, e.g. `writeln!`.

**Comment by huxi** [comment]
Currently suggested improved error message:
```console
error: expressions may not be used inside format strings
--> .\scratch\test.rs:3:37
  |
3 |     println!("hello {get_person()}");
  |                     ^^^^^^^^^^^^^^ expression is here
  |
= note: if you wanted to pass an expression as an argument to a formatting macro,
  try as a positional argument, e.g. println!("hello {}", get_person());
        or as a named argument, e.g. println!("hello {foo}", foo=get_person());
```

The message should also suggest using a local variable, e.g. `let foo = get_person(); println!("hello {foo}");`.
I would even argue that it should be the main suggestion since it promotes the same style as Field Init Shorthand. It also makes sense to promote implicit named arguments in this way.

**Comment by joshtriplett** [comment]
I've posted a stabilization PR at https://github.com/rust-lang/rust/pull/90473

**Comment by rben01** [comment]
Are there any plans to add support for something like Python's `=` suffix, which expands into `expr=value`? (And if not, could there be?) E.g., in Python, `x=1; print(f"{x+1=}")` prints `x+1=1`. (Julia also has a similar construct, the `@show` macro; `@show x+1;` → `x + 1 = 2`. Python's is a little more powerful, as it respects any normally-insignificant whitespace; `print(f"{x +1 = }")` → `x +1 = 2`.) This would be very handy when debugging, as `println!("{foo=}")` is a lot nicer than `println!("foo={foo}")`, and if support for expressions is added to `format_args_capture` then being able to do something like `println!("{f1(x, y, z)=}"` would be an even bigger win.

**Comment by programmerjake** [comment]
@rben01 
> Are there any plans to add support for something like Python's `=` suffix, which expands into `expr=value`?

The `dbg!` macro covers most of that need...

**Comment by rben01** [comment]
@programmerjake 
> The `dbg!` macro covers most of that need...

As I understand it, `dbg!(expr)` is the equivalent of Python's `print(f"{expr=}")`, but there isn't yet support for more general string interpolation. For instance, rust doesn't support `println!("Expected {expr1=}, got {expr2=}")`. Also, importantly, `dbg!` moves its input, whereas `println!("{expr1}")` only takes a reference .

**Comment by Mathspy** [comment]
> Also, importantly, `dbg!` moves its input, whereas `println!("{expr1}")` only takes a reference.

you can pass a reference to `dbg!(&a)`, the reason dbg! takes ownership is rather interesting, it’s because it returns what you pass to it too! So for example if you have some code like `let x = a + b + c` and you wanna find out the value of `b + c` you don’t have to temporarily add a variable and can instead do `let x = a + dbg!(b + c)` and the result of `b + c` needs to only implement `Debug` no clones or otherwise.
(On a side note this is a bit off topic and tracking issues are not designed for discussing features only for tracking them)

**Comment by bstrie** [comment]
Now that #90473 has been merged, can this be closed?

**Comment by AurevoirXavier** [comment]
Do we have a plan about the accepting expression `format!("{a.b()}")`

**Comment by davidhewitt** [comment]
The most recent discussion I've seen on that is at https://internals.rust-lang.org/t/how-to-allow-arbitrary-expressions-in-format-strings

**Comment by bionicles** [comment]
i'd love to use `println!("{x=}");` over `println!("x={x}");` because when the identifier for x is a_really_long_name then we have
`println!("{a_really_long_name=}");`
vs
`println!("a_really_long_name={a_really_long_name}");`
